### PR TITLE
gh-115999: Specialize loading attributes from modules in free-threaded builds

### PR DESCRIPTION
**Pull request by mpage**
This is the first in a series of PRs to specialize the `LOAD_ATTR` family of opcodes in free-threaded builds. PRs specializing `LOAD_ATTR` for instance and class receivers will follow. 

We use the same approach that was used for specialization of `LOAD_GLOBAL` in free-threaded builds:

`_CHECK_ATTR_MODULE` is renamed to `_CHECK_ATTR_MODULE_PUSH_KEYS`; it pushes the keys object for the following `_LOAD_ATTR_MODULE_FROM_KEYS` (nee `_LOAD_ATTR_MODULE`). This arrangement avoids having to recheck the keys version. 

`_LOAD_ATTR_MODULE` is renamed to `_LOAD_ATTR_MODULE_FROM_KEYS`; it loads the value from the keys object pushed by the preceding `_CHECK_ATTR_MODULE_PUSH_KEYS` at the cached index. 

A few other changes were necessary to support this arrangement:

* Added `POP_DEAD_INPUTS()` to the cases generator. It removes dead inputs from the stack and updates the stack pointer. This is used in `_LOAD_ATTR_MODULE_FROM_KEYS` to remove the keys object from the stack before deopting.
* Added a tier2-only implementation of `_LOAD_ATTR_MODULE`. This is used by the tier2 optimizer when we successfully erase a `_CHECK_ATTR_MODULE_PUSH_KEYS` but cannot convert the conjugate `_LOAD_ATTR_MODULE_FROM_KEYS` into a `LOAD_CONST`.
* Refactored the specialization logic to use the `specialize` and `unspecialize` helpers. This isn't strictly necessary, but it makes the implementation of gradually adding specialization support cleaner.
* Fixed `test_c_subclass_of_heap_ctype_with_del_modifying_dunder_class_only_decrefs_once` so that it works with deferred refcounting in the free-threaded build. Also refactored the test so that the evaluation of the argument expression to sys.getrefcount always uses deferred refcounting regardless of whether or not it's specialized.

### Specialization

A critical section on the module's dictionary is held during specialization. This prevents concurrent mutation of the dictionary and its keys version. `_PyDict_GetKeysVersionForCurrentState` is used to assign the keys version, ensuring that the keys object is freed using QSBR.

### UOPS

* Atomics are used where needed. Notably, loading the module's dictionary does not require an atomic: the `md_dict` field, but not the dictionary itself, is immutable.
* It's safe to access the keys object without holding the dictionary's per object lock: version assignment during specialization ensured it is freed using QSBR. A racing update will never store a different key at the cached index: once a unicode key has been stored in a keys object for a combined dictionary the offset that it was stored in will not be reused in free-threaded builds.

## Single-threaded Performance

- [Performance](https://github.com/facebookexperimental/free-threading-benchmarking/blob/main/results/bm-20241205-3.14.0a2%2B-eb8ce39) is neutral on the default build.
- [Performance](https://github.com/facebookexperimental/free-threading-benchmarking/tree/main/results/bm-20241205-3.14.0a2+-eb8ce39-NOGIL) is improved by ~1% on the free-threaded build.

## Scaling Benchmarks

This improves the scaling of calls to module functions (2.1x slower -> 13.9x faster), while other benchmarks appear unchanged.

```
Running benchmarks with 28 threads
                   Main          PR
object_cfunction    1.3x slower  1.4x slower
cmodule_function    1.5x slower  1.8x slower
mult_constant      11.8x faster  12.5x faster
generator          13.5x faster  12.6x faster
pymethod            2.0x slower  1.9x slower
pyfunction         13.7x faster  12.4x faster
module_function     2.1x slower  13.9x faster
load_string_const  13.6x faster  12.8x faster
load_tuple_const   13.6x faster  13.2x faster
create_pyobject    15.4x faster  15.9x faster
create_closure     14.4x faster  14.2x faster
create_dict        13.3x faster  13.5x faster
thread_local_read   3.6x slower  4.4x slower
```

* Issue: gh-115999

### CODE REVIEW

**Comment by mpage** [comment]
Refleaks build failures for the default build are preexisting. The free-threaded failure looks like its caused by this PR.

**Comment by mpage** [comment]
!buildbot nogil refleak

File: Python/bytecodes.c [hunk 1]
```diff
@@ -2157,33 +2157,43 @@ dummy_func(
...
-            assert(dict->ma_keys->dk_kind == DICT_KEYS_UNICODE);
-            assert(index < dict->ma_keys->dk_nentries);
-            PyDictUnicodeEntry *ep = DK_UNICODE_ENTRIES(dict->ma_keys) + index;
-            PyObject *attr_o = ep->me_value;
+            PyDictKeysObject *keys = FT_ATOMIC_LOAD_PTR_ACQUIRE(dict->ma_keys);
+            DEOPT_IF(FT_ATOMIC_LOAD_UINT32_RELAXED(keys->dk_version) != dict_version);
+            mod_keys = keys;
+        }
+
+        op(_LOAD_ATTR_MODULE_FROM_KEYS, (index/1, owner, mod_keys: PyDictKeysObject* -- attr, null if (oparg & 1))) {
```
Comment (colesbury) [inline Python/bytecodes.c]: fmt: `PyDictKeysObject *`

File: Lib/test/test_capi/test_misc.py [hunk 2]
```diff
@@ -665,19 +667,46 @@ def test_c_subclass_of_heap_ctype_with_del_modifying_dunder_class_only_decrefs_o
         del subclass_instance
 
         # Test that setting __class__ modified the reference counts of the types
+        #
```
Comment (colesbury) [inline Lib/test/test_capi/test_misc.py:670]: A bit of an aside, but I'm skeptical of the value of this test considering the increasing complexity. If we want to make sure we don't decref too many times, we should test that we don't crash. Depending on the exact refcounts feels too likely to break with implementation changes.

**Review by colesbury — APPROVED** [review]
LGTM

File: Lib/test/test_capi/test_misc.py (hunk 2 above)
Comment (mpage) [reply Lib/test/test_capi/test_misc.py:670]: Agreed. Checking that we don't crash sounds like a better approach to me. Filed https://github.com/python/cpython/issues/127881 to track that.

**Comment by mpage** [comment]
!buildbot AMD64 Android

**Comment by mpage** [comment]
Buildbot failures look unrelated to this PR:

* `AMD64 Arch Linux Asan PR`, `AMD64 Arch Linux Perf PR`, `AMD64 Windows11 Non-Debug PR`, `PPC64LE Fedora Stable LTO PR`, `PPC64LE Fedora Stable PR`, `PPC64LE RHEL8 LTO PR`    - All the same failure that exists on [main](https://buildbot.python.org/#/builders/146/builds/10224)
* AMD64 FreeBSD PR - Exists on [main](https://buildbot.python.org/#/builders/1223/builds/4778)
* `ARM64 Windows Non-Debug PR`, `ARM64 Windows PR` - Buildbots are brokens

**Pull request merged**